cmd/create/oidcprovider/cmd.go

```diff
@@ -0,0 +1,261 @@
+		reporter.Errorf("Cluster '%s' is %s and does not need additional configuration.",
+			clusterKey, cluster.State())
+		os.Exit(1)
+	}
+
+	if cluster.AWS().STS().RoleARN() == "" {
+		reporter.Errorf("Cluster '%s' is not an STS cluster.", clusterKey)
+		os.Exit(1)
+	}
+
+	mode := args.mode
+	if interactive.Enabled() {
+		mode, err = interactive.GetOption(interactive.Input{
+			Question: "Role creation mode",
+			Help:     cmd.Flags().Lookup("mode").Usage,
+			Default:  mode,
+			Options:  modes,
+			Required: true,
+		})
+		if err != nil {
+			reporter.Errorf("Expected a valid role creation mode: %s", err)
+			os.Exit(1)
+		}
+	}
+
+	switch mode {
+	case "auto":
+		reporter.Infof("Creating OIDC provider using '%s'", creator.ARN)
+		if !confirm.Confirm("create the OIDC provider for cluster '%s'", clusterKey) {
+			os.Exit(0)
+		}
+		err = createProvider(reporter, awsClient, cluster)
+		if err != nil {
+			reporter.Errorf("There was an error creating the OIDC provider: %s", err)
+			os.Exit(1)
+		}
+	case "manual":
+		reporter.Infof("Run the following commands to create the OIDC provider:\n")
+
+		commands, err := buildCommands(reporter, cluster)
+		if err != nil {
+			reporter.Errorf("There was an error building the list of resources: %s", err)
+			os.Exit(1)
+		}
+		fmt.Println(commands)
+	}
+}
+
+func createProvider(reporter *rprtr.Object, awsClient aws.Client, cluster *cmv1.Cluster) error {
+	oidcEndpointURL := cluster.AWS().STS().OIDCEndpointURL()
+
+	thumbprint, err := getThumbprint(oidcEndpointURL)
+	if err != nil {
+		return err
+	}
+	reporter.Debugf("Using thumbprint '%s'", thumbprint)
+
+	oidcProviderARN, err := awsClient.CreateOpenIDConnectProvider(oidcEndpointURL, thumbprint)
+	if err != nil {
+		return nil
+	}
+	reporter.Infof("Created OIDC provider with ARN '%s'", oidcProviderARN)
+
+	return nil
+}
+
+func buildCommands(reporter *rprtr.Object, cluster *cmv1.Cluster) (string, error) {
+	commands := []string{}
+
+	oidcEndpointURL := cluster.AWS().STS().OIDCEndpointURL()
+
+	thumbprint, err := getThumbprint(oidcEndpointURL)
+	if err != nil {
+		return "", err
+	}
+	reporter.Debugf("Using thumbprint '%s'", thumbprint)
+
+	createOpenIDConnectProvider := fmt.Sprintf("aws iam create-open-id-connect-provider \\\n"+
+		"\t--url %s \\\n"+
+		"\t--client-id-list %s %s \\\n"+
+		"\t--thumbprint-list %s",
+		oidcEndpointURL, aws.OIDCClientIDOpenShift, aws.OIDCClientIDSTSAWS, thumbprint)
+	commands = append(commands, createOpenIDConnectProvider)
+
+	return strings.Join(commands, "\n\n"), nil
+}
+
+func getThumbprint(oidcEndpointURL string) (string, error) {
+	connect, err := url.ParseRequestURI(oidcEndpointURL)
+	if err != nil {
+		return "", err
+	}
+
+	conn, err := tls.Dial("tcp", fmt.Sprintf("%s:443", connect.Host), nil)
+	if err != nil {
+		return "", err
+	}
+	defer conn.Close()
+
+	// Grab the last certificate in the chain
+	certChain := conn.ConnectionState().PeerCertificates
+	cert := certChain[len(certChain)-1]
```

Comment:
@vkareh can you think of a way to ensure that this is the CA certificate and not the intermediate certificate? I can't right now. Would be happy to mess around tomorrow and try to figure it out. Point being CA certs are less likely to expire or get rotated so thumbprints should be more stable.

Reply:
Fixed now. I'm looking for the CA, and if it's not there then it falls back to the last one in the chain.